From 39c94f54a2332f840ad5ae833a5420f14ebc6f3a Mon Sep 17 00:00:00 2001
From: Michael Dawson <midawson@redhat.com>
Date: Fri, 23 Feb 2024 13:43:56 +0100
Subject: [PATCH] Disable FIPS options

On RHEL, FIPS should be configured only on system level.
Additionally, the related options may cause segfault when used on RHEL.

This patch causes the option processing to end sooner
than the problematic code gets executed.
Additionally, the JS-level options to mess with FIPS settings
are similarly disabled.

Upstream report: https://github.com/nodejs/node/pull/48950
RHBZ: https://bugzilla.redhat.com/show_bug.cgi?id=2226726

Signed-off-by: rpm-build <rpm-build>
---
 lib/crypto.js             | 10 ++++++++++
 lib/internal/errors.js    |  6 ++++++
 src/crypto/crypto_util.cc |  4 ++++
 3 files changed, 20 insertions(+)

diff --git a/lib/crypto.js b/lib/crypto.js
index 3e720cd..523745a 100644
--- a/lib/crypto.js
+++ b/lib/crypto.js
@@ -36,6 +36,9 @@ const {
 assertCrypto();
 
 const {
+  // RHEL specific error
+  ERR_CRYPTO_FIPS_SYSTEM_CONTROLLED,
+
   ERR_CRYPTO_FIPS_FORCED,
   ERR_WORKER_UNSUPPORTED_OPERATION,
 } = require('internal/errors').codes;
@@ -267,6 +270,13 @@ function getFips() {
 }
 
 function setFips(val) {
+  // in RHEL FIPS enable/disable should only be done at system level
+  if (getFips() != val) {
+    throw new ERR_CRYPTO_FIPS_SYSTEM_CONTROLLED();
+  } else {
+    return;
+  }
+
   if (getOptionValue('--force-fips')) {
     if (val) return;
     throw new ERR_CRYPTO_FIPS_FORCED();
diff --git a/lib/internal/errors.js b/lib/internal/errors.js
index 4da0121..fc3a69c 100644
--- a/lib/internal/errors.js
+++ b/lib/internal/errors.js
@@ -1119,6 +1119,12 @@ module.exports = {
 //
 // Note: Node.js specific errors must begin with the prefix ERR_
 
+// insert RHEL specific erro
+E('ERR_CRYPTO_FIPS_SYSTEM_CONTROLLED',
+  'Cannot set FIPS mode. FIPS should be enabled/disabled at system level. See' +
+  'https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/security_hardening/assembly_installing-the-system-in-fips-mode_security-hardening for more details.\n',
+  Error);
+
 E('ERR_ACCESS_DENIED',
   function(msg, permission = '', resource = '') {
     this.permission = permission;
diff --git a/src/crypto/crypto_util.cc b/src/crypto/crypto_util.cc
index 19a0877..00d6623 100644
--- a/src/crypto/crypto_util.cc
+++ b/src/crypto/crypto_util.cc
@@ -466,6 +466,10 @@ std::optional<std::string> ProcessFipsOptions() {
   }
 #endif
 
+  // Fedora: we do not support using these options at all
+  // If we get here, just error out.
+  return "Using options related to FIPS is not recommended, configure FIPS in openssl instead. See https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/security_hardening/assembly_installing-the-system-in-fips-mode_security-hardening for more details.";
+
   CryptoErrorList errors{CryptoErrorList::Option::NONE};
   if (!ncrypto::setFipsEnabled(true, &errors)) {
     std::string error = "OpenSSL error when trying to enable FIPS";
-- 
2.56.0

